### PR TITLE
grpc client default in plaintext mode

### PR DESCRIPTION
**Pull request by wangxinyi7**
### Description

In this PR, we build a CLI GRPC client. It communicates with GRPC server in non-secure way.

- `grpc-config.go` will be responsible to assemble the client config and load env values into config
- `grpc-flags.go` will collect CLI flags information and load into the config
- `grpc-client.go` will instantiate client instance and expose:
  - connection: caller can close after the GRPC call
  - client: caller can use it talk with server
  - config: caller can retrieve some information when needed

**Usage:**

In CLI, the caller will do the following step to instantiate a client:
1. LoadGRPCConfig
2. MergeFlagsIntoGRPCConfig
3. NewGRPCClient

### PR Checklist

* [x] updated test coverage
* [ ] ~external facing docs updated~
* [ ] ~appropriate backport labels added~
* [ ] ~not a security concern~

### CODE REVIEW

File: command/resource/client/grpc-config.go [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+// SPDX-License-Identifier: BUSL-1.1
+
+package client
+
+import (
+	"os"
+)
+
+const (
+	// GRPCAddrEnvName defines an environment variable name which sets the gRPC
```
Comment (JadhavPoonam) [inline command/resource/client/grpc-config.go:11]: This comment is not accurate

File: command/resource/client/grpc-client_test.go [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+	"github.com/hashicorp/consul/agent"
+	"github.com/hashicorp/consul/internal/resource"
+	"github.com/hashicorp/consul/internal/resource/demo"
+	"github.com/hashicorp/consul/proto-public/pbresource"
+	"github.com/hashicorp/consul/proto/private/prototest"
+	"github.com/hashicorp/consul/sdk/testutil"
+	"github.com/hashicorp/consul/testrpc"
+)
+
+func TestResourceRead(t *testing.T) {
```
Comment (JadhavPoonam) [inline command/resource/client/grpc-client_test.go:25]: I am not sure you need all commands for the test here since we are only demonstrating the "dialing connection bit" so a single test with apply and read is enough for the client.

Perhaps what test coverage it needs is that shows env var `CONSUL_GRPC_ADDR` is not ignored.

File: command/resource/read/read.go [hunk 3]
```diff
@@ -86,10 +86,6 @@ func (c *cmd) Run(args []string) int {
 			return 1
 		}
 	} else {
-		if len(args) < 2 {
```
Comment (JadhavPoonam) [inline command/resource/read/read.go:89]: Why was this removed? 🤔

File: command/resource/client/grpc-config.go [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+	// address for consul connect envoy. Note this isn't actually used by the api
+	// client in this package but is defined here for consistency with all the
+	// other ENV names we use.
+	GRPCAddrEnvName = "CONSUL_GRPC_ADDR"
```
Comment (JadhavPoonam) [inline command/resource/client/grpc-config.go:15]: Are you intending to add other generic env vars such as `CONSUL_NAMESPACE` in follow-up PRs?

File: command/resource/client/grpc-flags.go [hunk 5]
```diff
@@ -0,0 +1,58 @@
+// Copyright (c) HashiCorp, Inc.
+// SPDX-License-Identifier: BUSL-1.1
+
+package client
```
Comment (JadhavPoonam) [inline command/resource/client/grpc-flags.go:4]: This client is pretty bare bones right now, I am guessing you will add connection pool management in the TLS PR?

**Review by JadhavPoonam — COMMENTED** [review]
This PR is a good first step for adding gRPC support.

However, WDYT about hooking up the client into the CLI and having an end-to-end command working since that may change the client considerably (💭 say something like a  `demo` command if you don't want to start with one of the existing commands).

File: command/resource/client/grpc-client.go [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+	}
+	return &GRPCClient{
+		Client: pbresource.NewResourceServiceClient(conn),
+		Config: config,
+		Conn:   conn,
+	}, nil
+}
+
+func dial(c *GRPCConfig) (*grpc.ClientConn, error) {
+	dialCtx := context.Background()
```
Comment (dhiaayachi) [inline command/resource/client/grpc-client.go:35]: I think we should have a way to cancel the context here. Maybe use a cancel context and keep the cancel channel in the Client?

File: command/resource/client/grpc-client_test.go (hunk 2 above)
Comment (dhiaayachi) [reply command/resource/client/grpc-client_test.go:25]: I agree with @JadhavPoonam. I think the smallest test possible in here is to create a tcp server and check that when you create the client it connect to the address you specified in the config. No need to use a test agent and no need to do any of the commands.

**Review by dhiaayachi — CHANGES_REQUESTED** [review]
Sorry for all the delay @wangxinyi7 😞 
I left few comments.

File: command/resource/read/read.go (hunk 3 above)
Comment (wangxinyi7) [reply command/resource/read/read.go:89]: The length check was wrapped inside of the function call
https://github.com/hashicorp/consul/blob/b85c9e77bed00d371b2274f8c0ac36e9184a4ef6/command/resource/read/read.go#L90C1-L90C1

https://github.com/hashicorp/consul/blob/b85c9e77bed00d371b2274f8c0ac36e9184a4ef6/command/resource/helper.go#L154

File: command/resource/client/grpc-config.go (hunk 1 above)
Comment (wangxinyi7) [reply command/resource/client/grpc-config.go:11]: good catch, updated

File: command/resource/client/grpc-client.go (hunk 6 above)
Comment (wangxinyi7) [reply command/resource/client/grpc-client.go:35]: I reviewed the situation and I don't think I need the context here so I'll dial the server directly.

File: command/resource/client/grpc-config.go (hunk 4 above)
Comment (wangxinyi7) [reply command/resource/client/grpc-config.go:15]: Yes, that's the plan, the file of `grpc-config.go` hosts all of the config variables.

File: command/resource/client/grpc-flags.go (hunk 5 above)
Comment (wangxinyi7) [reply command/resource/client/grpc-flags.go:4]: Correct, this file of `grpc-flags.go` hosts all of the flags related values.

File: command/resource/client/grpc-client_test.go (hunk 2 above)
Comment (wangxinyi7) [reply command/resource/client/grpc-client_test.go:25]: @JadhavPoonam Actually the var `CONSUL_GRPC_ADDR` is ignorable since the default value is `localhost:8502`

@dhiaayachi @JadhavPoonam Sounds good, I trimmed my the test suites to just cover successful connection.

**Comment by wangxinyi7** [comment]
> This PR is a good first step for adding gRPC support.
> 
> However, WDYT about hooking up the client into the CLI and having an end-to-end command working since that may change the client considerably (💭 say something like a `demo` command if you don't want to start with one of the existing commands).

The purpose of the scaffolding is to make the issue of the PR modular and easier to review in a manageable size instead of thousand lines PR.

But I could make a PR stack to flesh out the whole picture. This one is the base, [TLS](https://github.com/hashicorp/consul/pull/19680) is branched off from this one, and I could implement a CLI command based on the TLS one. Stay tuned.

File: command/resource/client/grpc-client_test.go (hunk 2 above)
Comment (JadhavPoonam) [reply command/resource/client/grpc-client_test.go:25]: Right.. so we can't really add a test unless a command is using this which takes `CONSUL_GRPC_ADDR` param

File: command/resource/read/read.go (hunk 3 above)
Comment (JadhavPoonam) [reply command/resource/read/read.go:89]: Ahh.. I see. Thank you for cleaning that up.

**Comment by JadhavPoonam** [comment]
> This PR is a good first step for adding gRPC support.
> 
> However, WDYT about hooking up the client into the CLI and having an end-to-end command working since that may change the client considerably (💭 say something like a `demo` command if you don't want to start with one of the existing commands).

Ok, sounds good to me.

**Review by JadhavPoonam — APPROVED** [review]
LGTM

**Review by dhiaayachi — APPROVED** [review]
LGTM! 🚀
